javascript/organizing_your_javascript_code/objects_and_object_constructors.md

```diff
@@ -59,45 +61,271 @@ const playerTwoMarker = "O";
+  console.log("Congratulations!");
+  console.log(winningPlayer.name + " is the winner!");
+}
+```
+
+Or, what if we aren’t making a 2 player game, but something more complicated such as an online shopping site with a large inventory? In that case, using objects to keep track of each particular item’s name, price, description and other things is the only way to go. You will continue to use and see objects used in this way throughout the curriculum.
+
+### Objects as a design pattern
+
+The grouping power of objects isn't just useful for organizing data- it's useful for organizing *functionality* as well! Using objects for this purpose is one of the core tenants of Object Oriented Programming (OOP).
+
+The inroductory paragraph for Object Oriented Programming on Wikipedia says this:
+
+> Object-oriented programming (OOP) is a programming paradigm based on the concept of objects, which can contain data and code: data in the form of fields (often known as attributes or properties), and code in the form of procedures (often known as methods). In OOP, computer programs are designed by making them out of objects that interact with one another.
+
+Essentially, what this means is that code can be organized into objects that contain not only data, but also **methods** (or functions on an object) that interact with that data.
+
+Nearly *anything* you can think about can be described as an object. To do so, all you have to do is ask yourself is "What properties (physical or conceptual) does my thing have?", and "How can I interact with it?". The properties or attributes of a *thing* are expressed as properties, and the ways you can interact with that thing are expressed as methods.
+
+Let's take an example of some thing- we'll choose a lightbulb. A lightbulb can have a color, and it can be in either an 'on' state, or an 'off' state. These might be expressed as properties of a lightbulb object:
+
+```js
+const lightbulb = {
+  lightColor: 'fluorescent white',  // this lightbulb is white,
+  lit: false                        // and is currently 'off'
+}
+```
+
+You may want to have the ability to switch a lightbulb from it's unlit state to it's lit state, or vice-versa. To do that, you might add a *method*.
+
+The easiest way to get started creating methods to interact with your objects might be combining Object Literal syntax with JavaScript's `this` keyword. The `this` keyword is used to refer to the object a particular method is called from.
+
+The following is an example of using the `this` keyword to add two methods to our object, `switchOn`, and `switchOff`:
+
+```js
+const lightbulb = {
+  lightColor: 'fluorescent white',
+  lit: false,
+
+  // shorthand syntax for adding methods to objects
+  switchOn() {
+    // return false to indicate the light was already on
+    if(this.lit === true) return false
+
+    // return true if the state of the light changed to be on
+    this.lit = true
+    return true
+  },
```

Comment:
This code would seem to violate the [command query separation](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation) principle. Potentially not a big deal, just wondering why you think this should be returning a value AND setting the `lit` property?

Reply:
It's just a decision I made when writing the code. Haven't exactly heard of that principle previously.

I've personally just found having some commands return some kind of status to be useful. Interesting that you picked this one out, but not the RPS object. The usefulness of this isn't as clear, though, and you may be right that there's not a huge reason to include the additional code.

I do see what that principle is saying and what it's trying to avoid though- especially when the opposite is true and you want to perform some query without modification to the object.